### ISSUE TITLE
How do you calculate the per-card cross entropy for Learned Belief?

### ISSUE THREAD

**Issue by ravihammond** [facebookresearch/off-belief-learning#14]
Hi Hengyuan,

I hope you're well! I have a question regarding your [Learned Belief Search paper](https://arxiv.org/pdf/2106.09086v1.pdf), Figure 4: _"Per-card cross entropy with the true hand for different belief in games played by BP"_. I have trained my own belief model, and would like to investigate its performance using the same loss calculation that you've used in this figure.

How exactly are you comparing the belief with the true hand? After investigating the [sample()](https://github.com/facebookresearch/off-belief-learning/blob/ce8e4fdbdb75aac317568c507fd801af51b286ca/pyhanabi/belief_model.py#L196) function from this OBL repo, I can see that for each card in the players hand, you're storing the list of probabilities in [this prob array](https://github.com/facebookresearch/off-belief-learning/blob/ce8e4fdbdb75aac317568c507fd801af51b286ca/pyhanabi/belief_model.py#L226), which is used for sampling. To calculate the cross-entropy loss, are you using this array?

To help me understand this better, would it be possible for me to see the code you used to generate the per-card cross entropy loss for this figure?

**Comment by hengyuan-hu** [comment]
Are you talking about learned belief, orange curve in Figure 4? If so, it is simply the cross-entropy loss (i.e. negative log likelihood) per card.
`nll_per_card` here https://github.com/facebookresearch/off-belief-learning/blob/ce8e4fdbdb75aac317568c507fd801af51b286ca/pyhanabi/belief_model.py#L68

**Comment by ravihammond** [comment]
Perfect, this is what I was looking for, thanks!

I notice that the only place you're calling the belief model's `loss()` function is [here](https://github.com/facebookresearch/off-belief-learning/blob/ce8e4fdbdb75aac317568c507fd801af51b286ca/pyhanabi/train_belief.py#L293), in the `train_belief.py` file. Here you're sampling a batch of games from the replay buffer to get the ground truth data for the xentropy loss calculation. When generating the data for the orange line in the previously mentioned figure 4 of the LBS paper, did you generate the ground truth data in the same way? (by sampling from the replay buffer), or did you generate the data some other way?

I assume the code for this is in a different repo (your private LBS repo)

**Comment by hengyuan-hu** [comment]
Yeah pretty much the same as how we generate the training data. Just remember to turn off exploration. We just generate a fixed amount of games in the replay buffer and run the loss function on the entire replay buffer, no more sampling. It is fairly easy to modify the code to do that.

**Comment by ravihammond** [comment]
Awesome, this is exactly what I was thinking of doing. Thanks for the information!
